proposals/0000-csi-haskell.rst

```diff
@@ -0,0 +1,275 @@
+    Main.smallestDiv (Divs.hs:10:17-29)
+    Main.main (Divs.hs:13:15-28)
+    Main.main (Divs.hs:13:8-29)
+    Main.CAF (<entire-module>)
+  Recently evaluated locations:
+    Divs.hs:4:25-4:26  ... = []
+    Divs.hs:4:16-4:21  |...,i == n,...=... (was matched)
+    repeats (11 times in window):
+      Divs.hs:4:9-7:28   Main:divs>go
+      Divs.hs:7:21-7:28  ... = go (i+1)
+      Divs.hs:5:19-5:21  ...else d i
+      Divs.hs:8:9-8:28   Main:divs>d
+      Divs.hs:5:16-7:28  ... = if d i...
+      Divs.hs:4:16-4:21  |...,i == n,...=... (not matched)
+    Divs.hs:4:9-7:28   Main:divs>go
+    Divs.hs:3:1-8:28   Main:divs
+   Previous expressions
+     Divs.hs:10:1-10:29  Main:smallestDiv
+     Divs.hs:13:1-13:29  Main:main
+
+Showing where the empty list in question originates.
+
+
+Proposed Change Specification
+-----------------------------
+
+We propose introducing the ``-fhpc-trace`` and the optional ``-fhpc-trace-length``
+flags, and extending the HPC instrumentation to collect a trace of recently
+evaluated locations when these flags are present.
+
+The traces track both when an expression starts being evaluated
+and when it stops being evaluated, as well as tracking the current
+*evaluation depth*, i.e. how many expressions we have started evaluating
+but not finished yet.
+
+These two in combination allows us to produce richer traces and error messages,
+such as the "Previous expressions" in the message above, which helps us
+understand what is currently being evaluated and what is still unresolved,
+effectively embedding some notion of the callstack into the trace itself.
+
+
+
+Proposed Library Change Specification
+-------------------------------------
+
+
+An alternative is to change the existing `error` function to include
+the evaluation trace whenever the flag is present.
+
+
+
+Effect and Interactions
+-----------------------
+
+Tools that parse runtime error messages directly will be impacted when the
+flag is present.
+
+
+Costs and Drawbacks
+-------------------
+
+Most of this proposal has already been implemented and evaluated. Maintenance
+is similarly minimal, and should not be more than the current burden of
+maintaining HPC in general.
+
+Novices will have to be made aware of the new flag, which will be covered in
+the users guide. Since the goal is to improve error messages, hopefully it will
+Make the language easier to learn.
+
+With the flag enabled, binary size, wall time and memory use of programs
+will be impacted. However, since we assume it will only be used during testing
+and development, the impact on end-users should be minimal.
+
+
+Backward Compatibility
+----------------------
+
+Existing programs should not change. The `.tix` files produced by HPC will
+include the additional tracing information, however, this will be accompanied
+by changes to the HPC parser to accommodate this extra information.
+
+Tools that consume this file directly will be impacted, those who use the
+HPC library will not.
```

Comment:
Could you give more details on how you envision the migration strategy for tools like https://github.com/8c6794b6/hpc-codecov or https://github.com/brandonchinn178/hpc-lcov#readme ?
Concretely I see the following problem: These tools are built with a fixed version of the hpc library, so they can either read the tix files which don't include the traces, or they can read the tix files which do include the traces. Currently they do work with tix files for all (viable) GHC versions. If the format of the tix file changes with GHC version `9.xx`, and a new version of the `hpc` library is released, then you always have to pick the specific version of `hpc-lcov` or `hpc-codecov` for the GHC version that you used to compile your binary with, and it is not possible to provide a latest version of these tools which works for all current GHC versions. 

In view of this I would prefer if we go with your alternative 5) and just provide an additional file format. I don't think that providing an additional data type and its serialization and deserialization logic would be too much maintenance burden. In view of your usecase I don't even think it would be necessary to provide a modified variant of the parser that is implemented in C in the RTS, since that parser is only used for the feature to initialize the tix data structures with information from a previous run.

Reply:
Good points! We have a modified `hpc` library as well that would have to be integrated into the current one. We have some `CPP` magic to make it work across versions, but a separate file format could be the way to go. One reason for the same file is that the `.tix` file is essentially just the contents of the RTS memory that keeps the HPC tix information, so we wanted it to match more exactly. But we can certainly introduce another file (`.trx` maybe?) that only keeps the trace). We won't submit to the committee itself until we have an amicable solution!

Reply:
I think that the core of your proposal is additive (i.e. it just adds a feature to GHC), and therefore doesn't necessitate any backwards incompatible changes. The 1:1 correspondence between internal data structure and tix file was introduced when HPC was first implemented, and while it would be a nice invariant to keep, I think it is better to give up this invariant and keep backwards compatibility instead, even if this means that the implementation has to be slightly more complex.
In light of the focus on stability in GHC, it would be sad if  your proposal gets bogged down in discussions about backwards compatibility. I don't think it should be too complicated to formulate the proposal in a way that promises that there are no backwards incompatible changes. You could just say that the semantics of `-fhpc` stays the same, and `-fhpc-trace` produces a `.trx` file, which is a clone of the  `.tix` file, but with the additional trace information.